overlay_new.py

```diff
@@ -0,0 +1,197 @@
+from abc import ABC, abstractmethod
+from time import time
+
+from canvas import Canvas, Colour
+from data import Data
+from overlay import Overlay
+from textwrap import wrap
+from typing import Callable, Tuple
+
+class Drawable(ABC):
+    @abstractmethod
+    def draw_base(self, canvas: Canvas):
+        pass
+    @abstractmethod
+    def draw_data(self, canvas: Canvas, data: Data):
+        pass
+
+class DataField(Drawable):
+    """ A simple data field containing a large value below a title in a smaller
+        font. Text is right-aligned. """
+
+    title_size = 0.8
+    data_size = 1.5
+
+    width, data_height = Canvas.get_text_dimensions("100.0", data_size)
+    _, title_height = Canvas.get_text_dimensions("KPH", title_size)
+
+    spacing = 10
+
+    height = title_height + data_height + spacing
+
+    def __init__(self, title: str, value_func: Callable[[Data], str], coordinate: Tuple[int, int]):
+        """ `coordinate` specifies the bottom-left coordinate of the data field. """
+        self.title = title
+        self.value_func = value_func
+        self.title_coord = (coordinate[0] + DataField.width, coordinate[1] - DataField.spacing - DataField.data_height)
+        self.data_coord = (coordinate[0] + DataField.width, coordinate[1])
+
+    def draw_base(self, canvas: Canvas):
+        canvas.draw_text(self.title, self.title_coord, DataField.title_size, Colour.white, "right")
+
+    def draw_data(self, canvas: Canvas, data: Data):
+        canvas.draw_text(self.value_func(data), self.data_coord, DataField.data_size, Colour.white, "right")
+
+class SpeedField(DataField):
+    """ A specialised version of DataField which prefers to display GPS speed,
+        but falls back to reed velocity if GPS is unavailable """
+
+    def __init__(self, coordinate: Tuple[int, int]):
+        value_func = lambda data: "{:.1f}".format(data["gps_speed"] or data["reed_velocity"])
+        super().__init__("", value_func, coordinate)
+
+    def draw_base(self, canvas: Canvas):
+        pass
```

Comment:
If this worked, then we could leave the method out altogether, because the implementation would default to the superclass implementation.

But, if we do this, the title gets drawn to the base layer rather than the data layer, which is not what we want in this specific case, as unlike other datafields, the title may change. This means when `draw_base` is called initially, `REED KPH` is written to the base layer (as no data has been received yet) but later when we have GPS data, `GPS KPH` is written to the data layer.

![image](https://user-images.githubusercontent.com/17876556/84472645-ebf08080-acca-11ea-8de0-16aa3328d5a7.png)

Reply:
ahhh I get you now. I do feel like the titles should be part of the base layer. In your scenario, it would be helpful if we had a way to re-render the base layer with the new title instead. Happy for this to be in another PR though.